### PR TITLE
wp-env: Use Simple Git instead of NodeGit

### LINKED ISSUE #26660: Migrate wp-env to use simple-git instead of nodegit
With the new LTS upgrade (14.15.0), the `nodegit` prebuilt binary is not yet available and requires being built from source. This is an extra complication to require everyone using wp-env to have all the dependencies installed.

This also highlighted that we probably don't need all the features of `nodegit` and could probably switch over to use `simple-git` that does have a requirement that git be installed locally, but that is safe tto assume and require for the project, since you need git to check out the project.

### PR DESCRIPTION
**Pull request by ockham**
## Description
Fixes #26660. Alternative to #28804.

~Since `simple-git` doesn't expose download progress for `git clone`, we lose the progress meters (i.e. they only show 0% at the start, and 100% once the download is done, but no in-between states).~

Edit (see https://github.com/WordPress/gutenberg/pull/28848#issuecomment-780793265): Oh, turns out the [latest version of `simple-git`](https://github.com/steveukx/git-js/releases/tag/v2.35.0) (which was just released) now includes a proper [progress handler](https://github.com/steveukx/git-js#progress-events) :partying_face:

## How has this been tested?

- Remove your local `~/.wp-env` or `~/wp-env` folder (warning, this removes your local WP dev install).
- Run `./packages/env/bin/wp-env start`. Verify that it successfully clones the `WordPress` and `theme-experiments` repos, and starts a local WP install successfully.
- Furthermore, verify that CI tests pass.

### CODE REVIEW

File: packages/env/lib/download-sources.js
Comment (noahtallen) [inline packages/env/lib/download-sources.js:119]: One thing I was curious about when I started putting this together: with a shallow clone, will git.checkout still work for refs which are older commits or tags?

Comment (ockham) [reply packages/env/lib/download-sources.js:119]: That's a great question!

Only one way to find out :wink:

Comment (ockham) [reply packages/env/lib/download-sources.js:119]: I just tried with

```diff
diff --git a/.wp-env.json b/.wp-env.json
index ccd1231aa2..4d67d71c82 100644
--- a/.wp-env.json
+++ b/.wp-env.json
@@ -3,7 +3,7 @@
        "plugins": [
                "."
        ],
-       "themes": [ "WordPress/theme-experiments/tt1-blocks" ],
+       "themes": [ "WordPress/theme-experiments/tt1-blocks#tt1-blocks@0.4.3" ],
        "env": {
                "tests": {
                        "mappings": {
```

(see https://github.com/WordPress/gutenberg/pull/28741#issuecomment-775320560), and it worked :tada:

**Comment by ockham** [comment]
Seems to work locally, and in CI :tada: 

I'm still getting a few `Git#then is deprecated after version 1.72 and will be removed in version 2.x` warnings in the console (that screw up the progress messages -- that's not properly working with `simple-git` anyway) even more :confused:

**Comment by paaljoachim** [comment]
Please let me know of any adjustments I will need to make to the Block Editor documentation for setting up a local WP dev environment: https://developer.wordpress.org/block-editor/tutorials/devenv/
Thanks!

**Comment by ockham** [comment]
> Please let me know of any adjustments I will need to make to the Block Editor documentation for setting up a local WP dev environment: https://developer.wordpress.org/block-editor/tutorials/devenv/
> Thanks!

Will do! I don't expect many user-facing changes, except we might lose the progress meter while wp-env is cloning git repos.

**Comment by ockham** [comment]
It seems that #28741 is no longer blocked by this PR. It'd still be nice to get it in eventually (once the `Git#then` warnings are sorted out), but I might deprioritize it a bit.

**Comment by noahtallen** [comment]
@ockham Looks like we have to require `simple-git/promise` for promise-based ops: https://github.com/steveukx/git-js/issues/406#issuecomment-559618286

Seeing if that fixes the `then` error

**Comment by noahtallen** [comment]
Or perhaps not: https://github.com/steveukx/git-js#recently-deprecated--altered-apis

**Comment by noahtallen** [comment]
Part of the problem might be that we have a different version of it in packages/wp-env than we do at the root 🤔

**Comment by noahtallen** [comment]
> Part of the problem might be that we have a different version of it in packages/wp-env than we do at the root 🤔

Yep, after addressing this, everything is working correctly. However, `simple-git` is used extensively in [lib/git.js](https://github.com/WordPress/gutenberg/blob/master/bin/plugin/lib/git.js), so I'm worried that it might not be compatible with the newest version.

**Comment by noahtallen** [comment]
Ok, so yeah, that definitely caused a lot of problems. I'm thinking we should do the update and fix the commands (probably not _too_ hard) in `lib/git.js`. 🤔 

Alternatively, we could stay on a legacy version, and only utilize legacy-version-compatible syntax for wp-env.

Otherwise, this is testing pretty well for me. 👍

**Comment by ockham** [comment]
Thanks for tracking this down, Noah! :tada: 

> Ok, so yeah, that definitely caused a lot of problems. I'm thinking we should do the update and fix the commands (probably not _too_ hard) in `lib/git.js`.  (Would have to be done in a separate PR though)

Yep :+1: On a sidenote, `lib/git.js` is solely used by release related tools. Some of those will be obsolete once the plugin build workflow has been moved into GitHub actions in its entirety (see #28138); there are however a few commands (such as the perf tests runner, or the `wp/trunk` branch merge) that might need them a little longer.

> Alternatively, we could stay on a legacy version, and only utilize legacy-version-compatible syntax for wp-env.

I'd rather use the new version, especially since I think we'll see `lib/git.js` used less and less (see above). I'm a bit on the fence whether we should wait for #28138 to land (in case this will make a few of the functions in `lib/git.js` obsolete -- and thus there'd be less code to migrate to the new version), but maybe I'll try to get a quick grasp of how much work it would be to update it so we aren't blocked by any other PRs.
 
> Otherwise, this is testing pretty well for me.

:partying_face:

**Comment by ockham** [comment]
Looking at the CI errors and https://github.com/steveukx/git-js#usage, I think the fix might be as simple as

```diff
diff --git a/bin/plugin/lib/git.js b/bin/plugin/lib/git.js
index b8b4242800..9af6288b99 100644
--- a/bin/plugin/lib/git.js
+++ b/bin/plugin/lib/git.js
@@ -1,7 +1,7 @@
 /**
  * External dependencies
  */
-const SimpleGit = require( 'simple-git/promise' );
+const SimpleGit = require( 'simple-git' );
 
 /**
  * Internal dependencies

```

**Comment by ockham** [comment]
I'm getting a ton of TypeScript errors that look like this now (see CI):

```
Error: bin/plugin/lib/git.js(20,20): error TS2349: This expression is not callable.
  Type 'typeof import("/home/runner/work/gutenberg/gutenberg/node_modules/simple-git/typings/index")' has no call signatures.
```

They happen on lines where I try to instantiate `simple-git` like this:

```js
const SimpleGit = require( 'simple-git' );
const git = SimpleGit(); // <-- error
```

@jsnajdr Any chance this is related to https://github.com/WordPress/gutenberg/pull/28879?

**Comment by noahtallen** [comment]
Do we need to convert wp-env to typescript? :trollface:  I'd love to do that actually, the jsdoc types are both important and a pain to work with.

**Comment by noahtallen** [comment]
@ockham I added `// @ts-nocheck` to `lib/git.js`. It does "fix" the problem, but is obviously not ideal.

I think part of the problem is that this file is written as a node script (and therefore uses `require()`), but typescript is expecting normal imports?

For example, `import SimpleGit from 'simple-git'` removes the type issues. (But I'm not sure that would actually _work_ in the context of node CLI here)

**Comment by jsnajdr** [comment]
> Any chance this is related to #28879?

@ockham That seems unlikely to me. #28879 mostly adds support for transpiling TypeScript files with Babel. It doesn't make any changes that would affect how `tsc` is run and configured.

**Comment by ockham** [comment]
Thanks @noahtallen and @jsnajdr!

The TS ignore seems like a good solution for now.

---

Static analysis checks have been complaining about an outdated package-lock. Rebasing to attempt a fix :crossed_fingers:

**Comment by ockham** [comment]
FWIW, I've briefly tried removing the `// @ts-nocheck` from `bin/plugin/lib/git.js` (by using an ESM import for `simple-git`, and ESM exports for the individual functions, see diff below).

<details>
<summary>Diff</summary>

```diff
diff --git a/bin/plugin/lib/git.js b/bin/plugin/lib/git.js
index 6bf62a0f8f..0f9c64ddbb 100644
--- a/bin/plugin/lib/git.js
+++ b/bin/plugin/lib/git.js
@@ -1,8 +1,7 @@
-// @ts-nocheck
 /**
  * External dependencies
  */
-const SimpleGit = require( 'simple-git' );
+import SimpleGit from 'simple-git';
 
 /**
  * Internal dependencies
@@ -16,7 +15,7 @@ const { getRandomTemporaryPath } = require( './utils' );
  *
  * @return {Promise<string>} Repository local Path
  */
-async function clone( repositoryUrl ) {
+export async function clone( repositoryUrl ) {
 	const gitWorkingDirectoryPath = getRandomTemporaryPath();
 	const simpleGit = SimpleGit();
 	await simpleGit.clone( repositoryUrl, gitWorkingDirectoryPath, [
@@ -35,7 +34,7 @@ async function clone( repositoryUrl ) {
  *
  * @return {Promise<string>} Commit Hash
  */
-async function commit( gitWorkingDirectoryPath, message, filesToAdd = [] ) {
+export async function commit( gitWorkingDirectoryPath, message, filesToAdd = [] ) {
 	const simpleGit = SimpleGit( gitWorkingDirectoryPath );
 	await simpleGit.add( filesToAdd );
 	const commitData = await simpleGit.commit( message );
@@ -50,7 +49,7 @@ async function commit( gitWorkingDirectoryPath, message, filesToAdd = [] ) {
  * @param {string} gitWorkingDirectoryPath Local repository path.
  * @param {string} branchName Branch Name
  */
-async function createLocalBranch( gitWorkingDirectoryPath, branchName ) {
+export async function createLocalBranch( gitWorkingDirectoryPath, branchName ) {
 	const simpleGit = SimpleGit( gitWorkingDirectoryPath );
 	await simpleGit.checkoutLocalBranch( branchName );
 }
@@ -61,7 +60,7 @@ async function createLocalBranch( gitWorkingDirectoryPath, branchName ) {
  * @param {string} gitWorkingDirectoryPath Local repository path.
  * @param {string} branchName Branch Name
  */
-async function checkoutRemoteBranch( gitWorkingDirectoryPath, branchName ) {
+export async function checkoutRemoteBranch( gitWorkingDirectoryPath, branchName ) {
 	const simpleGit = SimpleGit( gitWorkingDirectoryPath );
 	await simpleGit.fetch( 'origin', branchName );
 	await simpleGit.checkout( branchName );
@@ -73,7 +72,7 @@ async function checkoutRemoteBranch( gitWorkingDirectoryPath, branchName ) {
  * @param {string} gitWorkingDirectoryPath Local repository path.
  * @param {string} tagName Tag Name
  */
-async function createLocalTag( gitWorkingDirectoryPath, tagName ) {
+export async function createLocalTag( gitWorkingDirectoryPath, tagName ) {
 	const simpleGit = SimpleGit( gitWorkingDirectoryPath );
 	await simpleGit.addTag( tagName );
 }
@@ -84,7 +83,7 @@ async function createLocalTag( gitWorkingDirectoryPath, tagName ) {
  * @param {string} gitWorkingDirectoryPath Local repository path.
  * @param {string} branchName Branch Name
  */
-async function pushBranchToOrigin( gitWorkingDirectoryPath, branchName ) {
+export async function pushBranchToOrigin( gitWorkingDirectoryPath, branchName ) {
 	const simpleGit = SimpleGit( gitWorkingDirectoryPath );
 	await simpleGit.push( 'origin', branchName );
 }
@@ -94,7 +93,7 @@ async function pushBranchToOrigin( gitWorkingDirectoryPath, branchName ) {
  *
  * @param {string} gitWorkingDirectoryPath Local repository path.
  */
-async function pushTagsToOrigin( gitWorkingDirectoryPath ) {
+export async function pushTagsToOrigin( gitWorkingDirectoryPath ) {
 	const simpleGit = SimpleGit( gitWorkingDirectoryPath );
 	await simpleGit.pushTags( 'origin' );
 }
@@ -115,7 +114,7 @@ async function discardLocalChanges( gitWorkingDirectoryPath ) {
  * @param {string} gitWorkingDirectoryPath Local repository path.
  * @param {string} branchName Branch Name
  */
-async function resetLocalBranchAgainstOrigin(
+export async function resetLocalBranchAgainstOrigin(
 	gitWorkingDirectoryPath,
 	branchName
 ) {
@@ -131,7 +130,7 @@ async function resetLocalBranchAgainstOrigin(
  * @param {string} gitWorkingDirectoryPath Local repository path.
  * @param {string} commitHash Branch Name
  */
-async function cherrypickCommitIntoBranch(
+export async function cherrypickCommitIntoBranch(
 	gitWorkingDirectoryPath,
 	commitHash
 ) {
@@ -146,7 +145,7 @@ async function cherrypickCommitIntoBranch(
  * @param {string} gitWorkingDirectoryPath Local repository path.
  * @param {string} sourceBranchName Branch Name
  */
-async function replaceContentFromRemoteBranch(
+export async function replaceContentFromRemoteBranch(
 	gitWorkingDirectoryPath,
 	sourceBranchName
 ) {
@@ -159,17 +158,3 @@ async function replaceContentFromRemoteBranch(
 		'.',
 	] );
 }
-
-module.exports = {
-	clone,
-	commit,
-	checkoutRemoteBranch,
-	createLocalBranch,
-	createLocalTag,
-	pushBranchToOrigin,
-	pushTagsToOrigin,
-	discardLocalChanges,
-	resetLocalBranchAgainstOrigin,
-	cherrypickCommitIntoBranch,
-	replaceContentFromRemoteBranch,
-};
```
</details>

That _almost_ worked, except for a small issue with `git.reset()`.

```
bin/plugin/lib/git.js:108:25 - error TS2769: No overload matches this call.
  Overload 1 of 3, '(mode: ResetMode, options?: string[] | ResetOptions | undefined, callback?: SimpleGitTaskCallback<string, GitError> | undefined): Response<...>', gave the following error.
    Argument of type '"hard"' is not assignable to parameter of type 'ResetMode'.
  Overload 2 of 3, '(mode: ResetMode, callback?: SimpleGitTaskCallback<string, GitError> | undefined): Response<string>', gave the following error.
    Argument of type '"hard"' is not assignable to parameter of type 'ResetMode'.
  Overload 3 of 3, '(options?: string[] | ResetOptions | undefined, callback?: SimpleGitTaskCallback<string, GitError> | undefined): Response<string>', gave the following error.
    Argument of type '"hard"' is not assignable to parameter of type 'string[] | ResetOptions | undefined'.

108  await simpleGit.reset( 'hard' );
```

That could be fixed by a type assertion:

```
import SimpleGit, { ResetMode } from 'simple-git';
/* ... */
await simpleGit.reset( 'hard' as ResetMode );
```

However, if I try to build that, I'm reminded that type assertions are only allowed in actual `*.ts` files:

```
bin/plugin/lib/git.js:108:35 - error TS8016: Type assertion expressions can only be used in TypeScript files.

108  await simpleGit.reset( 'hard' as ResetMode );
```

At that point, I broke off the experiment.

I did file an issue against `simple-git` about requiring the type assertion: https://github.com/steveukx/git-js/issues/575

**Comment by ockham** [comment]
Since we'll be losing the progress meter (well, other than it showing 0% and then 100% once it's done cloning), I've also [asked upstream](https://github.com/steveukx/git-js/issues/527#issuecomment-778174851) if they could add the necessary features that might help us re-enable the progress meter later.

**Comment by ockham** [comment]
The static analysis check is still complaining about an outdated `package-lock`. I can't repro locally when using npm 6, so I guess it's due to npm 7. I'll try to unblock that...

**Comment by noahtallen** [comment]
Yeah, I also noticed that 🤔

**Comment by ockham** [comment]
> The static analysis check is still complaining about an outdated `package-lock`. I can't repro locally when using npm 6, so I guess it's due to npm 7. I'll try to unblock that...

FWIW, lots of fun here (and in PRs linked therein): https://github.com/WordPress/gutenberg/issues/28834

**Comment by noahtallen** [comment]
🙃

**Comment by ockham** [comment]
> > The static analysis check is still complaining about an outdated `package-lock`. I can't repro locally when using npm 6, so I guess it's due to npm 7. I'll try to unblock that...
> 
> FWIW, lots of fun here (and in PRs linked therein): #28834

Rebased, now that #28890 has been merged :crossed_fingers:

**Comment by noahtallen** [comment]
Looking good to me! How do we make sure that `git.js` will be ok?

File: packages/env/CHANGELOG.md
Comment (noahtallen) [inline packages/env/CHANGELOG.md:7]: Should we make this a breaking change? I think it could break things for folks who don't have a `git` binary (though that would be rare)

Comment (ockham) [reply packages/env/CHANGELOG.md:7]: Maybe. You're right that it is kind of a breaking change :thinking: (OTOH, who will use `wp-env` if they don't have `git` installed? :sweat_smile: )

But yeah, let's make it breaking.

**Comment by ockham** [comment]
My main remaining concern is that we lose the progress meter -- or at least, it becomes less meaningful, to the point of being confusing:

- For TT1 Blocks (which is small), the meter jumps to 100% pretty quickly.
- For the WordPress repo, it stays at 0% forever (while cloning the repo), giving the impression that something isn't working, before finally jumping to 100%.

I've tried using `simple-git`'s `outputHandler` (as [suggested here](https://github.com/steveukx/git-js/issues/529#issuecomment-733758118)), but since I don't know the total bytes we expect to receive, I can't turn those into percentages.

Maybe I'll give it another shot -- showing _some_ download progress (even if it's absolute numbers) might be better than not indicating any progress at all.

**Comment by ockham** [comment]
Oh, turns out the [latest version of `simple-git`](https://github.com/steveukx/git-js/releases/tag/v2.35.0) (which was released yesterday) now includes a proper [progress handler](https://github.com/steveukx/git-js#progress-events) :partying_face:

**Comment by ockham** [comment]
> Oh, turns out the [latest version of `simple-git`](https://github.com/steveukx/git-js/releases/tag/v2.35.0) (which was released yesterday) now includes a proper [progress handler](https://github.com/steveukx/git-js#progress-events)

I've upgraded to v2.35.0 and tried using the new progress handler in 37ccf2a, but haven't got it working (still only seeing either 0 or 100% of progress).

One idea I had is that the [RegEx that's looking for progress messages](https://github.com/steveukx/git-js/commit/5508bd4b10c7bb5233f93446931cdaa90ffeae4f#diff-3e408f15a975cc4cecd0de8fdfc86a92f1fd2202e3aa3d92d57f7bdc1f9e8fdfR18) isn't compatible with my German locale. I tried aliasing `git` to `LC_ALL=en git`, but I'm not sure that does the trick.

@noahtallen Can you check if it works for you?

**Review by noahtallen — APPROVED** [review]
Yeah, it appears to be working for me. I will note that I had to run `npx wp-env destroy` to really get the full effect. I think this might just be because the progress is mostly for clone, and when fetching the latest changes (e.g. if the repos already exist), maybe there's not much progress to indicate, or it just happens:

https://user-images.githubusercontent.com/6265975/108286949-df1c0a80-713e-11eb-8a87-2c09d6e91ab3.mov

**Comment by ockham** [comment]
> Yeah, it appears to be working for me. I will note that I had to run `npx wp-env destroy` to really get the full effect. I think this might just be because the progress is mostly for clone, and when fetching the latest changes (e.g. if the repos already exist), maybe there's not much progress to indicate, or it just happens:
> 
>  Screen.Recording.2021-02-17.at.4.38.39.PM.mov

Awesome, so happy to hear it works for you, Noah!

I filed an upstream bug report about this: https://github.com/steveukx/git-js/issues/579

Since releases seem to happen frequently for `simple-git`, let's maybe wait for another day or so before merging in case we can bump `simple-git` to a version that includes a fix for this.

**Comment by ockham** [comment]
Oh, seems like it's actually working for me now :thinking: :tada: 

Maybe I didn't properly clean out my `~/wp-env` last night. Anyway, going to merge! :partying_face:

**Comment by noahtallen** [comment]
Nice!

**Comment by mkaz** [comment]
@paaljoachim I don't think there is any change to the developer documentation for this PR. The issue previously was when we bumped the Node version there was not a pre-built binary for `nodegit` which was the previous module. The new library simple-git does not need a binary built so it simplifies matters.

**Comment by paaljoachim** [comment]
Thanks Marcus! A thank you to Bernie and Noah as well!
It is good to see that wp-env is being updated.

I gave the documentation a rest for a while before I get back to it again.

**Pull request merged**
